File: clojure/space-age/src/space_age.clj
(ns space-age)

(defn on-earth [age-in-s]
  (/ (float age-in-s) 31557600))

(defn on-mercury [age-in-s]
  (/ (on-earth age-in-s) 0.2408467))

(defn on-venus [age-in-s]
  (/ (on-earth age-in-s) 0.61519726))

(defn on-mars [age-in-s]
  (/ (on-earth age-in-s) 1.8808158))

(defn on-jupiter [age-in-s]
  (/ (on-earth age-in-s) 11.862615))

(defn on-saturn [age-in-s]
  (/ (on-earth age-in-s) 29.447498))

(defn on-uranus [age-in-s]
  (/ (on-earth age-in-s) 84.016846))

(defn on-neptune [age-in-s]
  (/ (on-earth age-in-s) 164.79132))

File: c/space-age/space_age.c
#include "space_age.h"

#include <math.h>

#define EARTH_YEAR_IN_SECONDS   31557600.0

#define ORBITAL_PERIOD_MERCURY  0.2408467
#define ORBITAL_PERIOD_VENUS    0.61519726
#define ORBITAL_PERIOD_EARTH    1.0
#define ORBITAL_PERIOD_MARS     1.8808158
#define ORBITAL_PERIOD_JUPITER  11.862615
#define ORBITAL_PERIOD_SATURN   29.447498
#define ORBITAL_PERIOD_URANUS   84.016846
#define ORBITAL_PERIOD_NEPTUNE  164.79132

#define INVALID_PLANET          -1.0

#define CASE_PLANET(planet) \
    case planet: \
        return ORBITAL_PERIOD_ ## planet

static float orbital_period(planet_t planet)
{
    switch (planet) {
        CASE_PLANET(MERCURY);
        CASE_PLANET(VENUS);
        CASE_PLANET(EARTH);
        CASE_PLANET(MARS);
        CASE_PLANET(JUPITER);
        CASE_PLANET(SATURN);
        CASE_PLANET(URANUS);
        CASE_PLANET(NEPTUNE);
    }

    return INVALID_PLANET;
}

float age(planet_t planet, int64_t seconds)
{
    float period = orbital_period(planet);
    return period != INVALID_PLANET
        ? seconds / EARTH_YEAR_IN_SECONDS / period
        : INVALID_PLANET;
}
